File: 67.add_binary/addBinary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
    var a_arr = a.split('');
    var b_arr = b.split('');
    var c_arr = [];
    var carry = 0;
    while(a_arr.length !==0 && b_arr.length !==0){
        var a_b = Number(a_arr.pop());
        var b_b = Number(b_arr.pop());
        var c = (a_b + b_b + carry) % 2;
        carry = ((a_b + b_b + carry) > 1) ? 1 : 0;
        console.log(carry);
        c_arr.unshift(c);
    }
    var remainArray = (a_arr.length === 0) ? b_arr : a_arr;
    console.log(remainArray);
    
    if(carry > 0){
        for(var i = remainArray.length - 1; i>=0; i--){
            if((Number(remainArray[i]) + carry) <= 1){
                remainArray[i] = Number(remainArray[i]) + carry;
                carry = 0;
                break;
            }
            remainArray[i] = 0;
        }
        if(carry > 0) remainArray.unshift(1);
    }
    var result = remainArray.concat(c_arr);
    console.log(result.join(""));
    return result.join("");
};

addBinary('0','1');

File: 67.add_binary/addBinary.py
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        if not a:
            return b
        if not b:
            return a;
        a_2 = int(a,2) # convert binary number to decimal
        b_2 = int(b,2)
        carry = 1
        while carry:
            carry = (a_2 + b_2) << 1
            a_2 = a_2 ^ b_2
            b_2 = carry
        return bin(a_2)[2:]